File: react-mvi-example/src/core/event-dispatcher.ts
import { Event } from 'src/core/event';
import { Observable, Subject } from 'rxjs';

export class EventDispatcher {
    private readonly eventSubject: Subject<Event>;
    public get eventStream(): Observable<Event> { return this.eventSubject; }

    constructor() {
        this.eventSubject = new Subject<Event>();
    }

    public dispatch(event: Event) {
        if (event) {
            console.log(`Event dispatched: ${event.eventType}`)
            this.eventSubject.next(event);
        }
    }
}

File: android-mvi-example/app/src/main/java/com/servicetitan/mviexample/core/EventDispatcher.kt
package com.servicetitan.mviexample.core

import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class EventDispatcher @Inject constructor() {
    private val eventSubject: PublishSubject<Event> = PublishSubject.create()
    val eventStream get(): Observable<Event> = eventSubject

    fun dispatch(event: Event) {
        eventSubject.onNext(event)
    }
}
